### PR TITLE
Simplify the way to get file names and add top_k_by_category

### PR DESCRIPTION
**Pull request by xiaoyuliu**
1. Path variable can return the file name without suffix by `path.stem`
2. Add check for whether `arg.stgcn_json_path` exists
3. Use `os.path.join` to avoid creating wrong paths
4. Add `top_k_by_category` to display evaluation accuracy results by category
5. Add `display_recall_precision` for each category

Please let me know your suggestion. Thank you.

### CODE REVIEW

File: st_gcn/feeder/feeder.py [hunk 1]
```diff
@@ -125,6 +125,44 @@ def top_k(self, score, top_k):
         hit_top_k = [l in rank[i, -top_k:] for i, l in enumerate(self.label)]
         return sum(hit_top_k) * 1.0 / len(hit_top_k)
 
+    def top_k_by_category(self, score, top_k):
```
Comment (yjxiong) [inline st_gcn/feeder/feeder.py]: Please write this as utility function and reuse them.

File: st_gcn/feeder/feeder.py [hunk 2]
```diff
@@ -125,6 +125,44 @@ def top_k(self, score, top_k):
...
+        accuracy_list = []
+        for hit_per_category in hit_top_k:
+            if hit_per_category:
+                accuracy_list.append(sum(hit_per_category) * 1.0 / len(hit_per_category))
+            else:
+                accuracy_list.append(0.0)
+        return accuracy_list
+
+    def calculate_recall_precision(self, score):
+        instance_num, class_num = score.shape
```
Comment (yjxiong) [inline st_gcn/feeder/feeder.py]: ditto

File: st_gcn/feeder/feeder_kinetics.py [hunk 3]
```diff
@@ -160,6 +160,48 @@ def top_k(self, score, top_k):
         hit_top_k = [l in rank[i, -top_k:] for i, l in enumerate(self.label)]
         return sum(hit_top_k) * 1.0 / len(hit_top_k)
 
+    def top_k_by_category(self, score, top_k):
+        assert (all(self.label >= 0))
```
Comment (yjxiong) [inline st_gcn/feeder/feeder_kinetics.py]: ditto

File: st_gcn/feeder/feeder_kinetics.py [hunk 4]
```diff
@@ -160,6 +160,48 @@ def top_k(self, score, top_k):
...
+        accuracy_list = []
+        for hit_per_category in hit_top_k:
+            if hit_per_category:
+                accuracy_list.append(sum(hit_per_category) * 1.0 / len(hit_per_category))
+            else:
+                accuracy_list.append(0.0)
+        return accuracy_list
+
+    def calculate_recall_precision(self, score):
+        assert (all(self.label >= 0))
```
Comment (yjxiong) [inline st_gcn/feeder/feeder_kinetics.py]: ditto

**Comment by yjxiong** [comment]
Thanks. Have you tested the training and testing? If the changes work properly I am going to merge them.

**Comment by xiaoyuliu** [comment]
Yes, I've tested on NTU-RGB-D dataset and also my custom datasets for training and testing.

**Comment by yjxiong** [comment]
Great. Did you get the expected accuracy? Would you please list the NTU-RGB accuracy for future reference?

**Comment by xiaoyuliu** [comment]
I didn't try to retrain the model from scratch, so not sure whether training can be reproduced. For testing yes, I got 80.65% (top 1) and 96.72% (top5).  The output is kind of long. 😄 

> Mean test loss of 258 batches: 0.781063414879.
Class1 Top1: 87.23%
Class2 Top1: 65.45%
Class3 Top1: 76.19%
Class4 Top1: 82.42%
Class5 Top1: 79.27%
Class6 Top1: 84.73%
Class7 Top1: 81.45%
Class8 Top1: 96.70%
Class9 Top1: 95.97%
Class10 Top1: 52.38%
Class11 Top1: 46.89%
Class12 Top1: 44.12%
Class13 Top1: 78.60%
Class14 Top1: 93.82%
Class15 Top1: 92.39%
Class16 Top1: 79.49%
Class17 Top1: 74.82%
Class18 Top1: 84.98%
Class19 Top1: 87.96%
Class20 Top1: 92.28%
Class21 Top1: 91.21%
Class22 Top1: 85.04%
Class23 Top1: 80.66%
Class24 Top1: 93.48%
Class25 Top1: 75.91%
Class26 Top1: 97.82%
Class27 Top1: 98.91%
Class28 Top1: 85.09%
Class29 Top1: 55.27%
Class30 Top1: 66.55%
Class31 Top1: 48.55%
Class32 Top1: 73.91%
Class33 Top1: 76.09%
Class34 Top1: 78.99%
Class35 Top1: 81.88%
Class36 Top1: 88.00%
Class37 Top1: 72.46%
Class38 Top1: 69.93%
Class39 Top1: 71.01%
Class40 Top1: 81.52%
Class41 Top1: 57.97%
Class42 Top1: 95.29%
Class43 Top1: 98.18%
Class44 Top1: 57.61%
Class45 Top1: 87.32%
Class46 Top1: 81.88%
Class47 Top1: 70.65%
Class48 Top1: 85.82%
Class49 Top1: 80.36%
Class50 Top1: 83.94%
Class51 Top1: 92.39%
Class52 Top1: 86.23%
Class53 Top1: 78.62%
Class54 Top1: 84.06%
Class55 Top1: 93.43%
Class56 Top1: 86.96%
Class57 Top1: 86.55%
Class58 Top1: 88.04%
Class59 Top1: 98.17%
Class60 Top1: 96.38%
Top1: 80.65%
Class1 Precision: 66.02%, Recall: 87.23%
Class2 Precision: 76.60%, Recall: 65.45%
Class3 Precision: 75.09%, Recall: 76.19%
Class4 Precision: 73.05%, Recall: 82.42%
Class5 Precision: 83.21%, Recall: 79.27%
Class6 Precision: 94.72%, Recall: 84.73%
Class7 Precision: 88.54%, Recall: 81.45%
Class8 Precision: 97.06%, Recall: 96.70%
Class9 Precision: 98.87%, Recall: 95.97%
Class10 Precision: 71.86%, Recall: 52.38%
Class11 Precision: 54.24%, Recall: 46.89%
Class12 Precision: 43.48%, Recall: 44.12%
Class13 Precision: 84.19%, Recall: 78.60%
Class14 Precision: 90.53%, Recall: 93.82%
Class15 Precision: 94.80%, Recall: 92.39%
Class16 Precision: 74.32%, Recall: 79.49%
Class17 Precision: 68.56%, Recall: 74.82%
Class18 Precision: 80.00%, Recall: 84.98%
Class19 Precision: 86.69%, Recall: 87.96%
Class20 Precision: 89.32%, Recall: 92.28%
Class21 Precision: 96.51%, Recall: 91.21%
Class22 Precision: 89.96%, Recall: 85.04%
Class23 Precision: 75.43%, Recall: 80.66%
Class24 Precision: 82.96%, Recall: 93.48%
Class25 Precision: 80.31%, Recall: 75.91%
Class26 Precision: 97.11%, Recall: 97.82%
Class27 Precision: 100.00%, Recall: 98.91%
Class28 Precision: 67.83%, Recall: 85.09%
Class29 Precision: 49.19%, Recall: 55.27%
Class30 Precision: 52.59%, Recall: 66.55%
Class31 Precision: 87.01%, Recall: 48.55%
Class32 Precision: 54.99%, Recall: 73.91%
Class33 Precision: 87.50%, Recall: 76.09%
Class34 Precision: 68.99%, Recall: 78.99%
Class35 Precision: 88.63%, Recall: 81.88%
Class36 Precision: 86.43%, Recall: 88.00%
Class37 Precision: 72.20%, Recall: 72.46%
Class38 Precision: 86.55%, Recall: 69.93%
Class39 Precision: 84.85%, Recall: 71.01%
Class40 Precision: 89.64%, Recall: 81.52%
Class41 Precision: 83.77%, Recall: 57.97%
Class42 Precision: 86.51%, Recall: 95.29%
Class43 Precision: 98.54%, Recall: 98.18%
Class44 Precision: 53.90%, Recall: 57.61%
Class45 Precision: 74.38%, Recall: 87.32%
Class46 Precision: 79.58%, Recall: 81.88%
Class47 Precision: 73.86%, Recall: 70.65%
Class48 Precision: 83.39%, Recall: 85.82%
Class49 Precision: 75.17%, Recall: 80.36%
Class50 Precision: 85.19%, Recall: 83.94%
Class51 Precision: 88.24%, Recall: 92.39%
Class52 Precision: 95.20%, Recall: 86.23%
Class53 Precision: 82.20%, Recall: 78.62%
Class54 Precision: 82.86%, Recall: 84.06%
Class55 Precision: 92.09%, Recall: 93.43%
Class56 Precision: 88.89%, Recall: 86.96%
Class57 Precision: 85.30%, Recall: 86.55%
Class58 Precision: 89.67%, Recall: 88.04%
Class59 Precision: 95.71%, Recall: 98.17%
Class60 Precision: 96.73%, Recall: 96.38%
Class1 Top5: 95.26%
Class2 Top5: 87.64%
Class3 Top5: 95.24%
Class4 Top5: 95.97%
Class5 Top5: 93.82%
Class6 Top5: 99.64%
Class7 Top5: 94.91%
Class8 Top5: 99.27%
Class9 Top5: 100.00%
Class10 Top5: 92.31%
Class11 Top5: 95.24%
Class12 Top5: 98.53%
Class13 Top5: 96.31%
Class14 Top5: 98.91%
Class15 Top5: 100.00%
Class16 Top5: 99.27%
Class17 Top5: 97.08%
Class18 Top5: 97.80%
Class19 Top5: 97.08%
Class20 Top5: 97.43%
Class21 Top5: 99.63%
Class22 Top5: 98.54%
Class23 Top5: 96.72%
Class24 Top5: 99.64%
Class25 Top5: 92.70%
Class26 Top5: 98.91%
Class27 Top5: 100.00%
Class28 Top5: 96.36%
Class29 Top5: 93.09%
Class30 Top5: 97.82%
Class31 Top5: 89.49%
Class32 Top5: 92.39%
Class33 Top5: 94.57%
Class34 Top5: 96.74%
Class35 Top5: 93.84%
Class36 Top5: 94.55%
Class37 Top5: 98.55%
Class38 Top5: 95.29%
Class39 Top5: 94.20%
Class40 Top5: 98.91%
Class41 Top5: 87.68%
Class42 Top5: 100.00%
Class43 Top5: 99.64%
Class44 Top5: 93.48%
Class45 Top5: 98.19%
Class46 Top5: 97.10%
Class47 Top5: 93.48%
Class48 Top5: 100.00%
Class49 Top5: 96.73%
Class50 Top5: 97.45%
Class51 Top5: 99.28%
Class52 Top5: 99.28%
Class53 Top5: 98.55%
Class54 Top5: 94.57%
Class55 Top5: 98.54%
Class56 Top5: 98.91%
Class57 Top5: 98.91%
Class58 Top5: 97.83%
Class59 Top5: 100.00%
Class60 Top5: 100.00%
Top5: 96.72%
Class1 Precision: 66.02%, Recall: 87.23%
Class2 Precision: 76.60%, Recall: 65.45%
Class3 Precision: 75.09%, Recall: 76.19%
Class4 Precision: 73.05%, Recall: 82.42%
Class5 Precision: 83.21%, Recall: 79.27%
Class6 Precision: 94.72%, Recall: 84.73%
Class7 Precision: 88.54%, Recall: 81.45%
Class8 Precision: 97.06%, Recall: 96.70%
Class9 Precision: 98.87%, Recall: 95.97%
Class10 Precision: 71.86%, Recall: 52.38%
Class11 Precision: 54.24%, Recall: 46.89%
Class12 Precision: 43.48%, Recall: 44.12%
Class13 Precision: 84.19%, Recall: 78.60%
Class14 Precision: 90.53%, Recall: 93.82%
Class15 Precision: 94.80%, Recall: 92.39%
Class16 Precision: 74.32%, Recall: 79.49%
Class17 Precision: 68.56%, Recall: 74.82%
Class18 Precision: 80.00%, Recall: 84.98%
Class19 Precision: 86.69%, Recall: 87.96%
Class20 Precision: 89.32%, Recall: 92.28%
Class21 Precision: 96.51%, Recall: 91.21%
Class22 Precision: 89.96%, Recall: 85.04%
Class23 Precision: 75.43%, Recall: 80.66%
Class24 Precision: 82.96%, Recall: 93.48%
Class25 Precision: 80.31%, Recall: 75.91%
Class26 Precision: 97.11%, Recall: 97.82%
Class27 Precision: 100.00%, Recall: 98.91%
Class28 Precision: 67.83%, Recall: 85.09%
Class29 Precision: 49.19%, Recall: 55.27%
Class30 Precision: 52.59%, Recall: 66.55%
Class31 Precision: 87.01%, Recall: 48.55%
Class32 Precision: 54.99%, Recall: 73.91%
Class33 Precision: 87.50%, Recall: 76.09%
Class34 Precision: 68.99%, Recall: 78.99%
Class35 Precision: 88.63%, Recall: 81.88%
Class36 Precision: 86.43%, Recall: 88.00%
Class37 Precision: 72.20%, Recall: 72.46%
Class38 Precision: 86.55%, Recall: 69.93%
Class39 Precision: 84.85%, Recall: 71.01%
Class40 Precision: 89.64%, Recall: 81.52%
Class41 Precision: 83.77%, Recall: 57.97%
Class42 Precision: 86.51%, Recall: 95.29%
Class43 Precision: 98.54%, Recall: 98.18%
Class44 Precision: 53.90%, Recall: 57.61%
Class45 Precision: 74.38%, Recall: 87.32%
Class46 Precision: 79.58%, Recall: 81.88%
Class47 Precision: 73.86%, Recall: 70.65%
Class48 Precision: 83.39%, Recall: 85.82%
Class49 Precision: 75.17%, Recall: 80.36%
Class50 Precision: 85.19%, Recall: 83.94%
Class51 Precision: 88.24%, Recall: 92.39%
Class52 Precision: 95.20%, Recall: 86.23%
Class53 Precision: 82.20%, Recall: 78.62%
Class54 Precision: 82.86%, Recall: 84.06%
Class55 Precision: 92.09%, Recall: 93.43%
Class56 Precision: 88.89%, Recall: 86.96%
Class57 Precision: 85.30%, Recall: 86.55%
Class58 Precision: 89.67%, Recall: 88.04%
Class59 Precision: 95.71%, Recall: 98.17%
Class60 Precision: 96.73%, Recall: 96.38%

**Comment by xiaoyuliu** [comment]
And I just notice there are duplicate precision/recall output for every `top_k`. Have modified the script

**Comment by yjxiong** [comment]
Looks good to me! Proceeding to merge.

**Comment by OpenMMLab-Assistant001** [comment]
Hi @xiaoyuliu！

First of all, we want to express our gratitude for your significant PR in the mmskeleton project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）

Thank you again for your contribution❤

**Pull request merged**